### PR TITLE
fix(gdn): compile SM12x CuteDSL kernels as sm_121a on DGX Spark

### PR DESCRIPTION
**Pull request by nv-tusharma**
## Summary
- Compile SM12x GDN CuteDSL kernels with a device-matched arch (`sm_120a` / `sm_121a`). Hardcoded `sm_120a` left DGX Spark (SM121) without a CuteDSL execution engine.
- Align GDN prefill test gates with `is_sm12x_supported` so SM121 is exercised.

Validated on NVIDIA GB10 (SM121, CUDA 13.0): representative GDN prefill/decode cases pass after the arch fix; b12x MoE suite was also green in the same sweep (`142 passed`).

Relates to #3170 (item 5).

## Test plan
- [x] `pytest tests/gdn/test_prefill_delta_rule.py -k 'test_prefill_kernel_basic and float16-64-seq_lens0-4-1-1-128'` on GB10 — 24 passed
- [x] Small GDN decode smoke (`basic_pretranspose` / `basic_nontranspose` B=1,4) on GB10 — 4 passed
- [x] `pytest tests/utils/test_jit_example.py::test_dump_logits --runxfail` still fails as expected on SM121
- [ ] CI / SM120 regression sanity if available

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Added dynamic SM12x GPU architecture targeting for Delta Rule kernel compilation.
  * Extended SM12x support across prefill and context-parallel Delta Rule operations.

* **Bug Fixes**
  * Improved SM12x PTX validation to prevent caching unsupported generated PTX.
  * Ensured compilation options are selected based on the actual tensor device for SM12x.

* **Tests**
  * Updated architecture gating in Delta Rule prefill tests to recognize SM12x (and adjust context-parallel skip logic accordingly).

### CODE REVIEW

File: flashinfer/gdn_kernels/delta_rule_dsl/custom_compile_cache.py [hunk 1]
```diff
@@ -2,14 +2,39 @@
...
 import cutlass.cute as cute
+import torch
 from cutlass.base_dsl.compiler import DumpDir
 
 
 _in_mem_compile_cache: dict = {}
 
 
+@lru_cache(maxsize=8)
+def sm12x_gpu_arch(device: str = "cuda") -> cute.GPUArch:
```
[the remark this answers is not in the document]
Comment (nv-tusharma) [reply flashinfer/gdn_kernels/delta_rule_dsl/custom_compile_cache.py]: Addressed in the latest commit: `sm12x_gpu_arch` / `sm12x_compile_options` now accept `str | torch.device` (normalized before the lru_cache).

File: flashinfer/gdn_kernels/delta_rule_dsl/custom_compile_cache.py [hunk 2]
```diff
@@ -2,14 +2,54 @@
...
+def sm12x_gpu_arch(device: str | torch.device = "cuda") -> cute.GPUArch:
+    """Return the CuteDSL GPUArch that can execute on this SM12x device.
+
+    CuteDSL treats mismatched arch-specific targets as cross-compilation and
+    produces no execution engine. SM120 must use ``sm_120a`` and SM121 (GB10 /
+    DGX Spark) must use ``sm_121a``; family targets like ``sm_120f`` do not run
+    on-device either.
+    """
+    d = torch.device(device)
+    return _sm12x_gpu_arch_cached(d.type, d.index)
```
[the remark this answers is not in the document]
Comment (nv-tusharma) [reply flashinfer/gdn_kernels/delta_rule_dsl/custom_compile_cache.py:36]: Good catch — fixed in the latest signed commit by resolving index-less `cuda` devices to `torch.cuda.current_device()` before the cache key is formed.

File: flashinfer/gdn_kernels/delta_rule_dsl/delta_rule_cp_sm120.py [hunk 3]
```diff
@@ -556,7 +560,7 @@ def cp_delta_rule_t_precompute_dsl_sm120(
         stream,
     )
     compiled = cached_compile(
-        kernel, *kernel_args, compile_options=(cute.GPUArch("sm_120a"),)
+        kernel, *kernel_args, compile_options=sm12x_compile_options()
```
Comment (kahyunnam) [inline flashinfer/gdn_kernels/delta_rule_dsl/delta_rule_cp_sm120.py]: Maybe we can just get the tensor device here? 

eg `compile_options=sm12x_compile_options(k.device)`

**Review by kahyunnam — COMMENTED** [review]
Why not sm120f for spark?

**Comment by nv-tusharma** [comment]
@kahyunnam  CuTeDSL’s runtime compatibility map lists only sm_121a for CC 12.1: https://github.com/NVIDIA/cutlass/blob/main/python/CuTeDSL/cutlass/base_dsl/runtime/cuda.py#L143
I tested sm_120f locally and the gdn kernel compiles, but invocation fails with `The compiled function does not have a valid execution engine`

File: flashinfer/gdn_kernels/delta_rule_dsl/delta_rule_cp_sm120.py (hunk 3 above)
Comment (nv-tusharma) [reply flashinfer/gdn_kernels/delta_rule_dsl/delta_rule_cp_sm120.py]: Updated all five compile sites to derive the target from their input tensor devices (`k.device`, `local_transfer.device`, or `q.device`). Validated with one non-CP prefill case and targeted tests covering all four CP compile paths on GB10.

**Review by kahyunnam — APPROVED** [review]
